### PR TITLE
Disclosure testing duplicate ids and keys

### PR DESCRIPTION
**Pull request by carlpulley-da**
Test that duplicate contract ids & keys are rejected
    
CHANGELOG_BEGIN
    
* Engine/speedy-level tests for explicit disclosure (#14227): test that duplicate contract ids & keys are rejected.
    
CHANGELOG_END

- [x] add duplicate disclosed contract ID testing to preprocessing tests
- [x] add duplicate disclosed contract key testing to preprocessing tests
- [x] rebase PR once https://github.com/digital-asset/daml/pull/14436 is merged
- [x] ensure PR passes in CI
- [x] ~~rewrite testing to use `preprocessDisclosedContracts`~~

Future PRs will add in additional testing as per https://github.com/digital-asset/daml/issues/14227

~~https://github.com/digital-asset/daml/pull/14576 will remove unused scenario error `PreprocessDisclosedContract` error code (as it is now unused).~~

### CODE REVIEW

**Review by cocreature — APPROVED** [review]
Tests look good to me thank you.

What do you think of testing this as a unit test for `buildDiscTable` instead of tryting to test everything at the more integration test style level you used here?

**Review by akrmn — APPROVED** [review]
Looks good to me, thank you

**Review by remyhaemmerle-da — APPROVED** [review]
Thanks.

**Comment by carlpulley-da** [comment]
> What do you think of testing this as a unit test for buildDiscTable instead of tryting to test everything at the more integration test style level you used here?

@cocreature after chatting with @remyhaemmerle-da, we can actually move the duplicate checks from `buildDiscTable` and place them in the more logical `Preprocessor#preprocessDisclosedContracts`. Doing this will also enable the removal of a lot of the existing scenario error code around disclosure errors (this will be managed in https://github.com/digital-asset/daml/pull/14576).

As a result, testing focus will now shift to this preprocessing method.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 1]
```diff
@@ -221,75 +221,24 @@ private[lf] object Speedy {
       err: String
   ) extends RuntimeException(err, null, true, false)
 
-  @throws[SErrorDamlException]
   private[speedy] def buildDiscTable(
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:224]: As this function can now operate assuming that `Preprocessor#preprocessDisclosuredContracts` has successfully ran, we no longer need any validation logic within this method.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExplicitDisclosureTest.scala [hunk 2]
```diff
@@ -8,11 +8,7 @@ import com.daml.lf.command.ContractMetadata
 import com.daml.lf.crypto.Hash
 import com.daml.lf.data.Ref.{IdString, Party}
 import com.daml.lf.data.{FrontStack, ImmArray, Ref, Struct, Time}
-import com.daml.lf.interpretation.Error.{
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExplicitDisclosureTest.scala:11]: All "disclosure preprocessing" testing has now been moved into the `PreprocessorSpec.scala` file.

**Comment by cocreature** [comment]
> @cocreature after chatting with @remyhaemmerle-da, we can actually move the duplicate checks from buildDiscTable and place them in the more logical Preprocessor#preprocessDisclosedContracts. Doing this will also enable the removal of a lot of the existing scenario error code around disclosure errors (this will be managed in https://github.com/digital-asset/daml/pull/14576).

`preprocessDisclosedContracts` won’t run for the scenario service. Where will the check live theres

**Comment by carlpulley-da** [comment]
> `preprocessDisclosedContracts` won’t run for the scenario service. Where will the check live there's

@cocreature that's a good point. If at all possible, I'd prefer not to make scenario service any more of a special service than it may currently be. As a result, my immediate thought for managing this issue would be to do something like (with analogous changes for `Speedy#fromScenarioExpr`):
```
    // located in Speedy.scala
    def fromScenarioSExpr(
        compiledPackages: CompiledPackages,
        scenario: SExpr,
        disclosedContracts: ImmArray[command.DisclosedContract] = ImmArray.Empty,
    )(implicit loggingContext: LoggingContext): Machine = {
      val commandPreprocessor =
        new CommandPreprocessor(
          pkgInterface = compiledPackages.pkgInterface,
          requireV1ContractIdSuffix = requireV1ContractIdSuffix,
        )

      Machine.fromPureSExpr(
        compiledPackages = compiledPackages,
        expr = SEApp(scenario, Array(SEValue.Token)),
        disclosedContracts = commandPreprocessor.unsafePreprocessDisclosedContracts(disclosedContracts),
      )
    }
```

@cocreature and @remyhaemmerle-da would doing something like this be acceptable or can either of you see additional issues with this approach?

**Comment by cocreature** [comment]
I'm not sure this works out. The scenario service is one level too deep. It should already be dealing with normalized SValues so the whole conversion done as part of preprocessing doesn’t make sense.

**Comment by cocreature** [comment]
You'll also end up reintroducing it to the scenario service errors pretty much regardless of which approach you take because it’s an error you can actually trigger.

I think I’m not sure what we’re gaining by moving the code to the preprocessor.

File: daml-lf/interpreter/BUILD.bazel [hunk 3]
```diff
@@ -47,6 +47,9 @@ da_scala_library(
 da_scala_library(
     name = "interpreter-test-lib",
     srcs = glob(["src/test/**/*Lib.scala"]),
+    scala_deps = [
```
Comment (carlpulley-da) [inline daml-lf/interpreter/BUILD.bazel:50]: Needed as some of the shared testing code defines scaliest matchers

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExplicitDisclosureLib.scala [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+import com.daml.lf.speedy.SValue.SContractId
+import com.daml.lf.transaction.{GlobalKey, GlobalKeyWithMaintainers, TransactionVersion, Versioned}
+import com.daml.lf.value.Value
+import com.daml.lf.value.Value.{ContractId, ContractInstance}
+import org.scalatest.matchers.{MatchResult, Matcher}
+import com.daml.lf.testing.parser.Implicits._
+
+/** Shared test data and functions for testing explicit disclosure.
+  */
+object ExplicitDisclosureLib {
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExplicitDisclosureLib.scala:22]: A copy and paste of the `ExplicitDisclosureTest` companion object

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExplicitDisclosureTest.scala [hunk 5]
```diff
@@ -41,15 +33,6 @@ class ExplicitDisclosureTest extends ExplicitDisclosureTestMethods {
         getOwner(disclosedCaveContract.argument.toUnnormalizedValue) shouldBe Some(disclosureParty)
       }
 
-      "disclosure preprocessing" - {
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExplicitDisclosureTest.scala:44]: Test deletions mark where code has now moved into `BuildDisclosureTableTest`

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExplicitDisclosureTest.scala [hunk 6]
```diff
@@ -414,362 +365,9 @@ class ExplicitDisclosureTest extends ExplicitDisclosureTestMethods {
   }
 }
 
-object ExplicitDisclosureTest {
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExplicitDisclosureTest.scala:417]: Majority of code here has now moved into `ExplicitDisclosureLib`

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/BuildDisclosureTableTest.scala [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        val error = intercept[SError.SErrorDamlException] {
+          Speedy.buildDiscTable(
+            ImmArray(disclosedHouseContract, disclosedHouseContractWithDuplicateKey),
+            pkg.pkgInterface,
+          )
+        }
+
+        error shouldBe SError.SErrorDamlException(
+          DisclosurePreprocessing(
+            DisclosurePreprocessing.DuplicateContractKeys(houseTemplateId)
```
Comment (cocreature) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/BuildDisclosureTableTest.scala:74]: Slightly out of the scope for this PR but why does this error not include the key hash?

**Review by cocreature — APPROVED** [review]
Thanks!

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/BuildDisclosureTableTest.scala (hunk 7 above)
Comment (carlpulley-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/BuildDisclosureTableTest.scala:74]: Ah, oversight by myself. Corrected this on PR https://github.com/digital-asset/daml/pull/14592

**Pull request merged**